src/Storage/Bucket.php

```diff
@@ -0,0 +1,350 @@
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Storage;
+
+use Google\Cloud\Storage\Acl;
+use Google\Cloud\Storage\Connection\ConnectionInterface;
+use Google\Cloud\Storage\Object;
+
+class Bucket
+{
+    /**
+     * @var ConnectionInterface
+     */
+    private $connection;
+
+    /**
+     * @var array
+     */
+    private $identity;
+
+    /**
+     * @var array
+     */
+    private $data;
+
+    /**
+     * @var Acl
+     */
+    private $acl;
+
+    /**
+     * @var Acl
+     */
+    private $defaultAcl;
+
+    /**
+     * @param ConnectionInterface $connection
+     * @param string $name
+     * @param array $data
+     */
+    public function __construct(ConnectionInterface $connection, $name, $data = null)
+    {
+        $this->connection = $connection;
+        $this->identity = ['bucket' => $name];
+        $this->data = $data;
+        $this->acl = new Acl($this->connection, 'bucketAccessControls', $this->identity);
+        $this->defaultAcl = new Acl($this->connection, 'defaultObjectAccessControls', $this->identity);
+    }
+
+    /**
+     * Configure ACL for this bucket.
+     *
+     * @return Acl
+     */
+    public function acl()
+    {
+        return $this->acl;
+    }
+
+    /**
+     * Configure default object ACL for this bucket.
+     *
+     * @return Acl
+     */
+    public function defaultAcl()
+    {
+        return $this->defaultAcl;
+    }
+
+    /**
+     * Check whether or not the bucket exists.
+     *
+     * Example:
+     * ```
+     * $bucket->exists();
+     * ```
+     *
+     * @return bool
+     */
+    public function exists()
+    {
+        try {
+            $this->connection->getBucket($this->identity + ['fields' => 'name']);
+        } catch (\Exception $ex) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Upload data from a file.
+     *
+     * Example:
+     * ```
+     * $bucket->uploadFromFile(
+     *     fopen('image.png', 'r'),
+     *     'image.png',
+     *     ['contentType' => 'image/jpeg']
+     * );
+     * ```
+     *
+     * @param mixed $body
+     * @param string $destination Name of where the file will be stored.
+     * @param array $options Configuration options.
+     * @return Object
+     */
+    public function uploadFromFile($body, $destination, $options = [])
+    {
+        $response = $this->connection->uploadObject($options + [
+            'bucket' => $this->name,
+            'body' => $body,
+            'name' => $destination
+        ]);
+
+        $objectIdentity = $this->identity + ['object' => $destination];
+
+        return new Object($this->connection, $objectIdentity, $response);
+    }
+
+    /**
+     * Upload
+     *
+     * Example:
+     * ```
+     * ```
+     *
+     * @param string $path Path to the file to be uploaded.
+     * @param array $options Configuration options.
+     * @return Object
+     */
+    public function uploadFromPath($path, $options = [])
```

Comment:
Just a note for when this is implemented, it would be nice to autodetect the mime type of the given file and use it to set the content-type metadata on the object.